src/TraceEvent/TraceLog.cs

```diff
@@ -153,10 +153,227 @@ public static TraceLog OpenOrConvert(string etlOrEtlxFilePath, TraceLogOptions o
+            if (session.m_kernelSession != null)
+            {
+                // Make sure both sources only dispatch one at a time by taking a lock during dispatch.
+                session.m_kernelSession.Source.lockObj = traceLog.realTimeQueue;
+                session.m_associatedWithTraceLog = true;                         // Indicate that it is OK to have the m_kernelSession.
+                session.Source.lockObj = traceLog.realTimeQueue;
+
+                // Set up the callbacks to the kernel session.
+                traceLog.rawKernelEventSource = session.m_kernelSession.Source;
+                traceLog.SetupCallbacks(traceLog.rawKernelEventSource);
+                traceLog.rawKernelEventSource.unhandledEventTemplate.traceEventSource = traceLog;       // Make everything point to the log as its source.
+                traceLog.rawKernelEventSource.AllEvents += traceLog.onAllEventsRealTime;
+            }
+
+            return traceLog.realTimeSource;
+        }
+
+        /// <summary>
+        /// From a EventPipeSession, create a real time TraceLog Event Source. Like an EventPipeEventSource a TraceLogEventSource
+        /// will deliver events in real time. However an TraceLogEventSource has an underlying Tracelog (which you can access with
+        /// the .Log Property) which lets you get at aggregated information (Processes, threads, images loaded, and perhaps most
+        /// importantly TraceEvent.CallStack() will work. Thus you can get real time stacks from events).
+        /// </summary>
+        /// <param name="rundownSession">
+        /// If given, the rundownSession is used to initialize module and method information and then the session is closed.
+        /// This only makes sense in realtime sessions when you need to resolve function names.
+        /// </param>
+        /// <example>
+        /// var client = new DiagnosticsClient(Process.GetCurrentProcess().Id);
+        /// using var rundownSession = client.StartEventPipeSession(
+        ///     new EventPipeProvider(ClrTraceEventParser.ProviderName, EventLevel.Informational, (long) ClrTraceEventParser.Keywords.Default),
+        ///     requestRundown: true
+        /// );
+        /// EventPipeProvider[] providers = new[]
+        /// {   new EventPipeProvider(ClrTraceEventParser.ProviderName, EventLevel.Informational, (long) ClrTraceEventParser.Keywords.Default),
+        ///     new EventPipeProvider(SampleProfilerTraceEventParser.ProviderName, EventLevel.Informational),
+        /// };
+        /// var session = client.StartEventPipeSession(providers, requestRundown: false);
+        /// var eventSource = TraceLog.CreateFromEventPipeSession(session, rundownSession);
+        /// eventSource.Process();
+        /// </example>
+        public static TraceLogEventSource CreateFromEventPipeSession(EventPipeSession session, EventPipeSession rundownSession = null)
```

Comment:
Yes, you're right - you can't re-use the session as far as I can tell.  Barring some change in EventPIpeSession or the DiagnosticsClient, I think we'll have to do something here that eventually changes.

It occurs to me that the rundown session needs to be configured properly.  I wonder if we should have another construct in TraceEvent that handles this.  We could do something like this:

Create an EventPipeRundownConfiguration class that has two static factories:
 - None (creates a configuration that does not trigger rundown)
 - Enable (takes a DiagnosticsClient and creates a rundown session)

This would allow us to control how rundown occurs later on, and would allow us to make changes to use StartRundown without users having to change their code.  It's not exactly what I was hoping for, but maybe OK.  What do you think?

Reply:
Sounds good, I've made the change